File: app_manager/appmanager/src/main/java/app/editors/manager/ui/fragments/main/MainPagerFragment.java
package app.editors.manager.ui.fragments.main;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import app.editors.manager.R;
import app.editors.manager.app.App;
import app.editors.manager.managers.tools.PreferenceTool;
import app.editors.manager.ui.activities.main.MainActivity;
import app.editors.manager.ui.fragments.base.BaseAppFragment;
import app.editors.manager.ui.views.pager.PagingViewPager;
import app.editors.manager.ui.views.pager.ViewPagerAdapter;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;

public class MainPagerFragment extends BaseAppFragment {

    public static final String TAG = MainPagerFragment.class.getSimpleName();
    private static final String TAG_POSITION = "TAG_POSITION";
    private static final String TAG_VISIBLE = "TAG_VISIBLE";
    private static final String TAG_SCROLL = "TAG_SCROLL";

    public final static int PAGE_DOCS_MY = 0;

    private final static int OFFSCREEN_COUNT = 5;

    @BindView(R.id.main_view_pager)
    protected PagingViewPager mMainViewPager;

    @Inject
    PreferenceTool mPreferenceTool;

    private Unbinder mUnbinder;
    private MainActivity mMainActivity;
    private ViewPagerAdapter mViewPagerAdapter;
    private boolean mIsScroll = true;
    private boolean mIsVisible = true;

    public static MainPagerFragment newInstance(final int position) {
        final MainPagerFragment mainPagerFragment = new MainPagerFragment();
        final Bundle bundle = new Bundle();
        bundle.putInt(TAG_POSITION, position);
        mainPagerFragment.setArguments(bundle);
        return mainPagerFragment;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        App.getApp().getAppComponent().inject(this);
        try {
            mMainActivity = (MainActivity) context;
        } catch (ClassCastException e) {
            throw new RuntimeException(MainPagerFragment.class.getSimpleName() + " - must implement - " +
                    MainActivity.class.getSimpleName());
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (mUnbinder != null) {
            mUnbinder.unbind();
        }
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        final View view = inflater.inflate(R.layout.fragment_main_pager, container, false);
        mUnbinder = ButterKnife.bind(this, view);
        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        init(savedInstanceState);
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        outState.putBoolean(TAG_VISIBLE, mIsVisible);
        outState.putBoolean(TAG_SCROLL, mIsScroll);
        super.onSaveInstanceState(outState);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mMainActivity.getTabLayout().setupWithViewPager(null);
        mMainActivity = null;
        mMainViewPager.setAdapter(null);
        if (getArguments() != null) {
            getArguments().clear();
        }
    }

    private void init(final Bundle savedInstanceState) {
        mMainActivity.showAppBarLayout(true);
        mMainActivity.setAppBarStates(false);
        mMainActivity.setNavigationButton(false);

        mViewPagerAdapter = new AdapterForPages(getChildFragmentManager(), getFragments());
        mMainViewPager.setOffscreenPageLimit(OFFSCREEN_COUNT);
        mMainViewPager.setAdapter(mViewPagerAdapter);
        mMainViewPager.addOnPageChangeListener(mViewPagerAdapter);
        mMainActivity.getTabLayout().setupWithViewPager(mMainViewPager, true);
        mMainActivity.setActionButtonShow(true);

        restoreStates(savedInstanceState);
        getPositionArg();

        mMainActivity.checkAccountInfo();
    }

    private void restoreStates(final Bundle savedInstanceState) {
        if (savedInstanceState != null) {
            if (savedInstanceState.containsKey(TAG_VISIBLE)) {
                mIsVisible = savedInstanceState.getBoolean(TAG_VISIBLE);
            }

            if (savedInstanceState.containsKey(TAG_SCROLL)) {
                mIsScroll = savedInstanceState.getBoolean(TAG_SCROLL);
            }
        }
    }

    private void getPositionArg() {
        final Bundle bundle = getArguments();
        if (bundle != null && bundle.containsKey(TAG_POSITION)) {
            mMainViewPager.post(() -> setPosition(bundle.getInt(TAG_POSITION)));
        }
    }

    /*
     * Get fragments
     * */
    private List<ViewPagerAdapter.Container> getFragments() {
        final List<ViewPagerAdapter.Container> pairs = new ArrayList<>();
        if (!mPreferenceTool.isPersonalPortal()) {
            if (!mPreferenceTool.getIsVisitor()) {
                pairs.addAll(getMyFragments());
            }
            pairs.addAll(getVisitorFragments());
        } else {
            pairs.addAll(getMyFragments());
        }
        pairs.add(new ViewPagerAdapter.Container(DocsTrashFragment.newInstance(), getString(R.string.main_pager_docs_trash)));
        return pairs;
    }

    private List<ViewPagerAdapter.Container> getVisitorFragments() {
        final List<ViewPagerAdapter.Container> pairs = new ArrayList<>();
        pairs.add(new ViewPagerAdapter.Container(DocsShareFragment.newInstance(),
                getString(R.string.main_pager_docs_share)));
        pairs.add(new ViewPagerAdapter.Container(DocsCommonFragment.newInstance(),
                getString(R.string.main_pager_docs_common)));
        if (!mPreferenceTool.isProjectDisable()) {
            pairs.add(new ViewPagerAdapter.Container(DocsProjectsFragment.newInstance(),
                    getString(R.string.main_pager_docs_projects)));
        }
        pairs.add(new ViewPagerAdapter.Container(DocsFavoritesFragment.newInstance(), getString(R.string.main_pager_docs_favorites)));
        return pairs;
    }

    private List<ViewPagerAdapter.Container> getMyFragments() {
        final List<ViewPagerAdapter.Container> pairs = new ArrayList<>();
        pairs.add(new ViewPagerAdapter.Container(DocsMyFragment.newInstance(),
                getString(R.string.main_pager_docs_my)));
        return pairs;
    }

    public void showActionDialog() {
        ((DocsBaseFragment) getActiveFragment()).showActionDialog();
    }

    boolean isActivePage(final Fragment fragment) {
        return mViewPagerAdapter.isActiveFragment(mMainViewPager, fragment);
    }

    private Fragment getActiveFragment() {
        return mViewPagerAdapter.getActiveFragment(mMainViewPager);
    }

    /*
     * Set states
     * */

    public void setPosition(int position) {
        position = position - getOffsetPosition();
        if (position >= PAGE_DOCS_MY && position < mViewPagerAdapter.getCount()) {
            mMainViewPager.setCurrentItem(position, true);
        }
    }

    private int getOffsetPosition() {
        return mPreferenceTool.getIsVisitor() ? 1 : 0;
    }

    void setScrollViewPager(final boolean isScroll) {
        mIsScroll = isScroll;
        mMainViewPager.setPaging(mIsScroll);
    }

    void setToolbarState(final boolean isRoot) {
        mIsVisible = isRoot;
        mMainActivity.setAppBarStates(mIsVisible);
    }

    void setExpandToolbar() {
        mMainActivity.expandToolBar();
    }

    void setVisibilityActionButton(final boolean isShow) {
        mMainActivity.setActionButtonShow(isShow);
    }

    public int getPosition() {
        return mViewPagerAdapter.getSelectedPage();
    }

    public void setAccountEnable(boolean isEnable) {
        mMainActivity.setToolbarAccount(isEnable);
    }

    public void disableProjectModule() {
        if (mViewPagerAdapter != null){
            mViewPagerAdapter.removeFragment(mViewPagerAdapter.getCount()-1);
        }
    }

    public void setVisibleTabs(boolean isVisible) {
        mMainActivity.setAppBarStates(isVisible);
        mMainActivity.setNavigationButton(true);
    }

    /*
     * Adapter and page change listener
     * */
    private class AdapterForPages extends ViewPagerAdapter {

        AdapterForPages(FragmentManager manager, List<Container> fragmentList) {
            super(manager, fragmentList);
        }

        @Override
        public void onPageSelected(int position) {
            super.onPageSelected(position);
            mMainActivity.expandFloatingActionButton();
            mMainActivity.setPagePosition(position);
            ((DocsCloudFragment) getActiveFragment(mMainViewPager)).onScrollPage();
        }
    }

}

File: app_manager/appmanager/src/main/java/app/editors/manager/ui/fragments/main/MainPagerFragment.kt
package app.editors.manager.ui.fragments.main

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.StringRes
import androidx.core.view.isVisible
import androidx.core.view.updatePadding
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import app.documents.core.model.cloud.isDocSpace
import app.documents.core.network.common.contracts.ApiContract
import app.documents.core.network.manager.models.explorer.Explorer
import app.editors.manager.R
import app.editors.manager.app.App
import app.editors.manager.app.accountOnline
import app.editors.manager.app.appComponent
import app.editors.manager.databinding.FragmentMainPagerBinding
import app.editors.manager.managers.tools.PreferenceTool
import app.editors.manager.mvp.models.models.OpenDataModel
import app.editors.manager.mvp.presenters.main.MainPagerPresenter
import app.editors.manager.mvp.views.main.MainPagerView
import app.editors.manager.ui.activities.main.ActionButtonFragment
import app.editors.manager.ui.activities.main.IMainActivity
import app.editors.manager.ui.activities.main.MainActivity
import app.editors.manager.ui.fragments.base.BaseAppFragment
import app.editors.manager.ui.views.custom.PlaceholderViews
import app.editors.manager.ui.views.pager.ViewPagerAdapter
import app.editors.manager.ui.views.pager.ViewPagerAdapter.Container
import lib.toolkit.base.managers.utils.UiUtils
import lib.toolkit.base.managers.utils.clearIntent
import moxy.presenter.InjectPresenter

interface IMainPagerFragment {

    fun setPagerPosition(sectionType: Int, onPageChanged: () -> Unit = {})
    fun setToolbarInfo(title: String?, drawable: Int? = null)
}

class MainPagerFragment : BaseAppFragment(), ActionButtonFragment, MainPagerView, View.OnClickListener,
    IMainPagerFragment, PlaceholderViews.OnClickListener {

    companion object {

        val TAG: String = MainPagerFragment::class.java.simpleName

        private const val TAG_SELECTED_PAGE = "TAG_SELECTED_PAGE"
        private const val TAG_VISIBLE = "TAG_VISIBLE"
        private const val TAG_SCROLL = "TAG_SCROLL"
        private const val TAG_TITLES = "TAG_TITLES"
        private const val TAG_TYPE = "TAG_TYPE"
        private const val TAG_PERSONAL_END = "TAG_PERSONAL_END"
        private const val TAG_PAYMENT_REQUIRED = "TAG_PAYMENT_REQUIRED"
        private const val TAG_CONNECTION = "TAG_CONNECTION"
        private const val OFFSCREEN_COUNT = 5

        fun newInstance(): MainPagerFragment {
            return MainPagerFragment()
        }
    }

    @InjectPresenter
    lateinit var presenter: MainPagerPresenter

    private var adapter: AdapterForPages? = null
    private var activity: IMainActivity? = null
    private var isScroll = true
    private var isVisibleRoot = false
    private var selectedPage = 0

    private var viewBinding: FragmentMainPagerBinding? = null
    private var placeholderViews: PlaceholderViews? = null

    private var preferenceTool: PreferenceTool? = null

    private var tabTile: ArrayList<String>? = null
    private var type: ArrayList<Int>? = null

    override fun onAttach(context: Context) {
        super.onAttach(context)
        try {
            if (context is IMainActivity) {
                activity = context
            }
            preferenceTool = requireContext().appComponent.preference
        } catch (e: ClassCastException) {
            throw RuntimeException(
                MainPagerFragment::class.java.simpleName + " - must implement - " +
                        MainActivity::class.java.simpleName
            )
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        viewBinding = FragmentMainPagerBinding.inflate(inflater, container, false)
        return viewBinding?.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        init(savedInstanceState)
        if (App.getApp().showPersonalPortalMigration &&
            context?.accountOnline?.isPersonal() == true
        ) {
            activity?.showPersonalMigrationFragment()
        }
    }

    private var isFirstResume = true

    override fun onResume() {
        super.onResume()
        if (isFirstResume) {
            isFirstResume = false
        } else {
            viewBinding?.mainViewPager?.post {
                activeFragment?.onResume()
            }
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putBoolean(TAG_VISIBLE, isVisibleRoot)
        outState.putBoolean(TAG_SCROLL, isScroll)
        outState.putBoolean(
            TAG_PERSONAL_END,
            placeholderViews?.type == PlaceholderViews.Type.PERSONAL_PORTAL_END
        )
        outState.putBoolean(
            TAG_PAYMENT_REQUIRED,
            placeholderViews?.type == PlaceholderViews.Type.PAYMENT_REQUIRED
        )
        outState.putBoolean(
            TAG_CONNECTION,
            placeholderViews?.type == PlaceholderViews.Type.CONNECTION
        )
        outState.putInt(TAG_SELECTED_PAGE, selectedPage)
        outState.putStringArrayList(TAG_TITLES, tabTile)
        outState.putIntegerArrayList(TAG_TYPE, type)
        super.onSaveInstanceState(outState)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        activity = null
        viewBinding = null
    }

    private fun init(savedInstanceState: Bundle?) {
        supportActionBar?.title = ""
        placeholderViews = PlaceholderViews(viewBinding?.placeholderLayout?.root)
        if (savedInstanceState != null) {
            restoreStates(savedInstanceState)
        } else {
            placeholderViews?.setTemplatePlaceholder(PlaceholderViews.Type.LOAD)
            presenter.getState()
        }
    }

    private fun restoreStates(savedInstanceState: Bundle) {
        if (savedInstanceState.containsKey(TAG_VISIBLE)) {
            isVisibleRoot = savedInstanceState.getBoolean(TAG_VISIBLE)
        }
        if (savedInstanceState.containsKey(TAG_SCROLL)) {
            isScroll = savedInstanceState.getBoolean(TAG_SCROLL)
        }
        if (savedInstanceState.containsKey(TAG_SELECTED_PAGE)) {
            selectedPage = savedInstanceState.getInt(TAG_SELECTED_PAGE)
        }
        if (savedInstanceState.containsKey(TAG_PERSONAL_END)) {
            if (savedInstanceState.getBoolean(TAG_PERSONAL_END)) {
                placeholderViews?.setTemplatePlaceholder(PlaceholderViews.Type.PERSONAL_PORTAL_END)
            }
        }
        if (savedInstanceState.containsKey(TAG_PAYMENT_REQUIRED)) {
            if (savedInstanceState.getBoolean(TAG_PAYMENT_REQUIRED)) {
                placeholderViews?.setTemplatePlaceholder(PlaceholderViews.Type.PAYMENT_REQUIRED)
            }
        }
        if (savedInstanceState.containsKey(TAG_CONNECTION)) {
            if (savedInstanceState.getBoolean(TAG_CONNECTION)) {
                placeholderViews?.setTemplatePlaceholder(PlaceholderViews.Type.CONNECTION, ::onRetryClick)
            }
        }
        if (savedInstanceState.containsKey(TAG_TITLES) && savedInstanceState.containsKey(TAG_TYPE)) {
            tabTile = savedInstanceState.getStringArrayList(TAG_TITLES)
            type = savedInstanceState.getIntegerArrayList(TAG_TYPE)

            val fragments: MutableList<MainPagerContainer> = mutableListOf()
            childFragmentManager.fragments.forEachIndexed { index, fragment ->
                if (fragment is DocsBaseFragment) {
                    fragments.add(MainPagerContainer(fragment, tabTile?.get(index) ?: "", type?.get(index) ?: 0))
                }
            }
            setAdapter(fragments, true)
        } else {
            presenter.getState()
        }
    }

    fun isActivePage(fragment: Fragment?): Boolean {
        return adapter?.isActiveFragment(viewBinding?.mainViewPager, fragment) == true
    }

    fun setScrollViewPager(isScroll: Boolean) {
        this.isScroll = isScroll
        viewBinding?.mainViewPager?.isPaging = this.isScroll
    }

    fun setToolbarState(isRoot: Boolean, hideToolbarInfo: Boolean) {
        if (!isVisible) return
        isVisibleRoot = isRoot
        activity?.setAppBarStates(isVisibleRoot, hideToolbarInfo)
        viewBinding?.let { binding ->
            binding.appBarTabs.isVisible = isVisibleRoot
            if (!isTablet) {
                binding.mainViewPager.updatePadding(
                    bottom = if (isVisibleRoot) binding.appBarTabs.height else 0
                )
            }
        }
    }

    fun setExpandToolbar() {
        //        mMainActivity.expandToolBar();
    }

    fun setVisibilityActionButton(isShow: Boolean) {
        if (!isVisible) return
        activity?.showActionButton(isShow)
    }


    fun setAccountEnable(isEnable: Boolean) {
        if (!isVisible) return
        activity?.showAccount(isEnable)
    }

    fun setVisibleTabs(isVisible: Boolean) {
        activity?.setAppBarStates(isVisible)
    }

    override fun showActionDialog() {
        (activeFragment as? DocsBaseFragment)?.showActionDialog()
    }

    override fun onRender(sections: List<Explorer>?) {
        type = arrayListOf()
        tabTile = arrayListOf()
        sections?.let {
            val fragments = sections.mapNotNull { section ->
                type?.add(section.current.rootFolderType)
                when (val folderType = section.current.rootFolderType) {
                    ApiContract.SectionType.CLOUD_PRIVATE_ROOM -> null
                    else -> {
                        tabTile?.add(getTabTitle(folderType))
                        MainPagerContainer(
                            fragment = when (folderType) {
                                ApiContract.SectionType.CLOUD_TRASH,
                                ApiContract.SectionType.CLOUD_ARCHIVE_ROOM -> {
                                    DocsTrashFragment.newInstance(folderType, section.current.id)
                                }

                                ApiContract.SectionType.CLOUD_VIRTUAL_ROOM -> {
                                    DocsRoomFragment.newInstance(folderType, section.current.id)
                                }

                                else -> {
                                    DocsCloudFragment.newInstance(folderType, section.current.id)
                                }
                            },
                            title = getTabTitle(folderType),
                            sectionType = folderType
                        )
                    }
                }
            }
            setAdapter(fragments)
        }
    }

    override fun onFinishRequest() {
        if (placeholderViews?.type == PlaceholderViews.Type.PAYMENT_REQUIRED) {
            activity?.showActionButton(false)
        } else {
            placeholderViews?.setTemplatePlaceholder(PlaceholderViews.Type.NONE)
        }
    }

    override fun onError(message: String?) {
        if (message == getString(R.string.errors_client_payment_required)) {
            activity?.showActionButton(false)
            placeholderViews?.setTemplatePlaceholder(PlaceholderViews.Type.PAYMENT_REQUIRED) {
                context?.accountOnline?.portal?.urlWithScheme?.let(::showUrlInBrowser)
            }
        } else if (message == getString(R.string.errors_connection_error)) {
            placeholderViews?.setTemplatePlaceholder(PlaceholderViews.Type.CONNECTION, ::onRetryClick)
        } else {
            message?.let { showSnackBar(it) }
            (requireActivity() as? MainActivity)?.onUnauthorized(message)
        }
    }

    override fun onRetryClick() {
        placeholderViews?.setTemplatePlaceholder(PlaceholderViews.Type.LOAD)
        presenter.getState()
    }

    override fun onError(@StringRes res: Int) {
        showSnackBar(getString(res))
        requireActivity().intent.data = null
    }

    private fun setAdapter(fragments: List<MainPagerContainer>, isRestore: Boolean = false) {
        if (fragments.isEmpty()) {
            viewBinding?.appBarTabs?.isVisible = false
            return
        }

        adapter = AdapterForPages(childFragmentManager, fragments)
        viewBinding?.mainViewPager?.offscreenPageLimit = OFFSCREEN_COUNT
        viewBinding?.mainViewPager?.adapter = adapter
        adapter?.let {
            viewBinding?.mainViewPager?.addOnPageChangeListener(it)
        }
        viewBinding?.appBarTabs?.setupWithViewPager(viewBinding?.mainViewPager, true)
        setToolbarState(true, true)
        if (isRestore) {
            viewBinding?.mainViewPager?.currentItem = selectedPage
        } else {
            if (context?.accountOnline.isDocSpace) {
                viewBinding?.mainViewPager?.post {
                    viewBinding?.mainViewPager?.currentItem =
                        fragments.indexOf(fragments.find { it.mFragment is DocsRoomFragment })
                }
            }
        }
    }

    override fun onSwitchAccount(data: OpenDataModel) {
        UiUtils.showQuestionDialog(
            context = requireContext(),
            title = getString(R.string.switch_account_title),
            description = getString(R.string.switch_account_description, data.portal),
            acceptListener = {
                activity?.showAccountsActivity()
            },
            cancelListener = {
                presenter.onRemoveFileData()
                requireActivity().intent.clearIntent()
            },
            acceptTitle = getString(R.string.switch_account_open_project_file)
        )
    }

    override fun onPersonalPortalEnd() {
        placeholderViews?.setTemplatePlaceholder(PlaceholderViews.Type.PERSONAL_PORTAL_END)
    }

    fun isRoot(): Boolean {
        return if (activeFragment != null && activeFragment is DocsCloudFragment) {
            (activeFragment as DocsCloudFragment).isRoot
        } else {
            false
        }
    }

    val position: Int?
        get() = adapter?.selectedPage

    private val activeFragment: Fragment?
        get() = runCatching { adapter?.getActiveFragment(viewBinding?.mainViewPager) }.getOrNull()

    override fun setToolbarInfo(title: String?, drawable: Int?) {
        viewBinding?.infoLayout?.root?.isVisible = title != null
        viewBinding?.infoLayout?.infoText?.text = title
        viewBinding?.infoLayout?.infoText?.setCompoundDrawablesRelativeWithIntrinsicBounds(drawable ?: 0, 0, 0, 0)
    }

    override fun onClick(view: View?) {
        activity?.onSwitchAccount()
    }

    private fun getTabTitle(tab: Int): String =
        when (tab) {
            ApiContract.SectionType.CLOUD_COMMON -> requireContext().getString(R.string.main_pager_docs_common)
            ApiContract.SectionType.CLOUD_USER -> requireContext().getString(R.string.main_pager_docs_my)
            ApiContract.SectionType.CLOUD_FAVORITES -> requireContext().getString(R.string.main_pager_docs_favorites)
            ApiContract.SectionType.CLOUD_SHARE -> requireContext().getString(R.string.main_pager_docs_share)
            ApiContract.SectionType.CLOUD_TRASH -> requireContext().getString(R.string.main_pager_docs_trash)
            ApiContract.SectionType.CLOUD_PROJECTS -> requireContext().getString(R.string.main_pager_docs_projects)
            ApiContract.SectionType.CLOUD_VIRTUAL_ROOM -> requireContext().getString(R.string.main_pager_docs_virtual_room)
            ApiContract.SectionType.CLOUD_ARCHIVE_ROOM -> requireContext().getString(R.string.main_pager_docs_archive_room)
            ApiContract.SectionType.CLOUD_RECENT -> requireContext().getString(R.string.fragment_recent_title)
            else -> ""
        }

    override fun setPagerPosition(
        sectionType: Int,
        onPageChanged: () -> Unit
    ) {
        adapter?.let { adapter ->
            val index = adapter.fragmentList.indexOfFirst { it.sectionType == sectionType }
            if (index > -1) {
                viewBinding?.mainViewPager?.let { pager ->
                    pager.currentItem = index
                    pager.postDelayed(onPageChanged, 500)
                }
            }
        }
    }

    /*
     * Adapter and page change listener
     * */
    private data class MainPagerContainer(
        val fragment: Fragment,
        val title: String,
        val sectionType: Int
    ) : Container(fragment, title)

    private inner class AdapterForPages(
        manager: FragmentManager,
        val fragmentList: List<MainPagerContainer>
    ) : ViewPagerAdapter(manager, fragmentList) {

        override fun onPageSelected(position: Int) {
            super.onPageSelected(position)
            activity?.showActionButton(false)
            this@MainPagerFragment.selectedPage = selectedPage
            (getActiveFragment(viewBinding?.mainViewPager) as DocsCloudFragment).onScrollPage()
        }
    }
}